gitpod-io/gitpod
# User Namespace support in workspaces - Part I

This PR lays the foundation for using user namespaces to better isolate workspaces - and give them more capabilities in the process. The four main contributions of this PR are:
- **make workspace CRI introduced with ws-sync's FWB available in common-go**: with the development of full workspace backup, ws-sync introduced an abstraction on top of the actual container runtime used on the node. We called this abstraction the "container runtime interface", much like Kubernetes does. Currently, this CRI only supports `containerd` and its API is likely to change in the future. ws-manager-node also interacts with the container runtime on the node. Now that this interaction intensifies it was an opportune moment to integrate the workspace CRI here as well. For that to happen, I've pulled the CRI out of ws-sync and moved it to `common-go/cri`.
- **generalising ws-manager-node's resourcegov's dispatch and integrating the CRI**: ws-manager-node's resourcegov provides dynamic resource limiting (beyond what the Kubernetes CRI can do) by directly modifying a container's `cgroup`. To this end, we listen on Kubernetes for new pods and then talk to containerd (now the CRI abstraction) to wait for the container to become available. ws-manager-node calls this "wait for workspace and its container" mechanism _dispatch_. With the uidmapping canary that follows, this dispatch mechanism is needed outside of the resource governor. To facilitate that, I've moved the dispatch to its own package (`dispatch`) and integrated `common-go/cri`.
- **adding uid/gid mapping support to supervisor using a ws-manager-node backed canary**: establishing a [UID/GID mapping in a Linux user namespace](https://www.man7.org/linux/man-pages/man7/user_namespaces.7.html) requires `CAP_SYS_ADMIN`, `CAP_SETUID` and `CAP_SETGID` in the userns's parent. Especially `CAP_SYS_ADMIN` is not something we want to give the workspace container on the node. Hence, we have ws-manager-node execute the privileged operations required to establish the mapping. To this end, ws-manager-node:
  - connects to a workspace's supervisor (using the dispatch mechanism described before) and registers a service (canary) that can be used to trigger a new mapping
  - can translate a PID from the container's PID namespace to the node's PID space by inspecting the container's child process tree
  - validates the incoming UID/GID mapping request
  - writes the mapping to `/proc/PID/[u|g]id_map`
- **make user namespaces work by relaxing pod security**: processes inside the user namespace get a [full set of capabilities](https://github.com/torvalds/linux/blob/38696e33e2bdf91cbbc7a2291dc6db862d9dfd42/kernel/user_namespace.c#L47-L52), unless [`no_new_privs`](https://man7.org/linux/man-pages/man2/prctl.2.html) is set in which case `execve` does not pass on those capabilities (see [here](https://man7.org/linux/man-pages/man7/capabilities.7.html), "Transformation of capabilities during execve()" - thanks @meysholdt for finding the explanation for execve's behaviour). For mounts to work inside the user namespace, we also need to disable seccomp and apparmour for now. This is clearly a dangerous state of affairs, and the reason that the new `user_namespace` feature flag requires the `privileged-ws` permission. The `user_namespace` feature flag also hostPath mounts the node's `/dev/net/tun` device into the container, s.t. we can create tap network devices in the user namespace and use [slirp4netns](https://github.com/rootless-containers/slirp4netns).

Further, this PR makes all subsequent dev-staging preview environments use the Ubuntu node pool rather than the COS one. This allows us to use `shiftfs` and enables unprivileged `overlayfs` mounts.

This PR also fixes a minor issue in ws-manager, where it would try and apply a feature flag multiple times if it was specified multiple times in the `StartWorkspace` request.

## Impact on production
Production impact should be fairly minimal, as most security relevant behaviour is hidden behind the permission-guarded feature flag. Things to watch out for/discuss are:
- should we make the uidmapping canary available to workspaces started with the feature flag only? At the moment all workspaces get this service.
- are the UID mapping rules sufficient to be secure? I.e. can you spot a bug or misconfig that would result in insecure mappings?
- did the refactoring of the CRI and dispatch break anything, first and foremost the resource governor?

## How to test
1. log in to create your gitpod user: http://cw-workspace-usernamespace.staging.gitpod-dev.com/workspaces/
2. give yourself the `privileged-ws` permission and `user_namespace` feature flag:
  ```SQL
  UPDATE d_b_user 
  SET
    rolesOrPermissions="[\"privileged-ws\"]",
    featureFlags = "{\"permanentWSFeatureFlags\":[\"user_namespace\"]}" 
  WHERE
    name = "csweichel"
  ;
  ```
3. open a workspace, e.g. http://cw-workspace-usernamespace.staging.gitpod-dev.com/#https://github.com/csweichel/test-repo/tree/newuidmap
4. manually spawn a user namespace:
  terminal a:
  ```bash
  unshare -U bash
  echo $$ > /tmp/pid
  ```
  in a different terminal b:
  ```bash
  $HOME/bin/newuidmap $(cat /tmp/pid) 0 33333 1 1 100000 65000
  $HOME/bin/newgidmap $(cat /tmp/pid) 0 33333 1 1 100000 65000
  ```
  back in term a check which user you are:
  ```bash
  bash
  whoami
  id
  cat /proc/self/status | grep Cap
  ```
  Notice the full set of capabilities (use `capsh --decode` to understand what that means).
5. spawn a user namespace using rootlesskit:
  ```bash
  rootlesskit --copy-up /etc --net=slirp4netns bash
  ```
When trying the user namespaces, have a `top` running to observe that you are not actually `root`.
